File: lib/matching.rb
# typed: strict
require 'sorbet-runtime'
require './lib/csv_parser'
require './lib/preferences'
require 'descriptive_statistics'


class Matching
  extend T::Sig

  sig {params(person: Person2025).returns(T::Array[Person2025])}
  def self.multiplicity(person)
    if person.max_num_mentees <= 1
      return [person]
    end

    return (1..person.max_num_mentees).map{|i| person.with(id: "#{person.id}#{i}")}
  end

  sig {params(people: T::Array[Person2025]).void}
  def self.match(people)

    # To handle one mentor with multiple mentees, we create one Person object
    # per relationship of them matching (e.g. 3 mentees means 3 person objects)
    mentors = people.select(&:is_mentor).flat_map{|person| multiplicity(person)}
    mentees = people.select(&:is_mentee)

    mentees_to_preferences = Preferences.compute_mentee_to_mentor_preferences(mentees: mentees, mentors: mentors)
    mentees_with_no_preferences = mentees_to_preferences.select {|_, preferences| preferences.empty?}.keys
    if !mentees_with_no_preferences.empty?
      puts "Filtering out #{mentees_with_no_preferences.size} mentees with no preferences:\n#{mentees_with_no_preferences}"
      mentees_to_preferences = mentees_to_preferences.reject {|_, preferences| preferences.empty?}
    end
    puts "Num mentees: #{people.select(&:is_mentee).group_by(&:email).keys.size}"

    mentors_to_preferences = Preferences.compute_mentor_to_mentee_preferences(mentees: mentees, mentors: mentors)
    mentors_with_no_preferences = mentors_to_preferences.select {|_, preferences| preferences.empty?}.keys
    if !mentors_with_no_preferences.empty?
      puts "Filtering out #{mentors_with_no_preferences.size} mentors with no preferences:\n"
      mentors_to_preferences = mentors_to_preferences.reject {|_, preferences| preferences.empty?}
    end
    puts "Num mentors: #{people.select(&:is_mentor).group_by(&:email).keys.size}"
    puts "Num mentor slots (accounting for multiple mentees per mentor): #{mentors_to_preferences.keys.size}"

    matched_mentors_to_mentees = gale_shapley(proposers: mentees_to_preferences, acceptors: mentors_to_preferences)

    puts("\n\n")
    puts("*************RESULTS*************\n\n")

    compute_match_statistics(
      matched_mentors_to_mentees: matched_mentors_to_mentees,
      mentees: people.select(&:is_mentee), 
      mentors: mentors.select(&:is_mentor), 
      mentees_to_preferences: mentees_to_preferences, 
      mentors_to_preferences: mentors_to_preferences, 
    )

    puts("Mentees -> Mentors:\n\n", matched_mentors_to_mentees.map{|mentor, mentee| "#{mentee.name};#{mentee.email};#{mentor.name};#{mentor.email}"}.sort.join("\n"))
  end

  sig do
    params(
      matched_mentors_to_mentees: T::Hash[Person2025, Person2025],
      mentees: T::Array[Person2025],
      mentees_to_preferences: T::Hash[Person2025, T::Array[Person2025]],
      mentors: T::Array[Person2025],
      mentors_to_preferences: T::Hash[Person2025, T::Array[Person2025]],
    ).void
  end
  private_class_method def self.compute_match_statistics(
    matched_mentors_to_mentees:,
    mentees:, 
    mentees_to_preferences:, 
    mentors:, 
    mentors_to_preferences: 
  )
    matched_mentees = matched_mentors_to_mentees.values.map(&:email).uniq
    matched_mentors = matched_mentors_to_mentees.keys.map(&:email).uniq

    unique_mentees = mentees.map(&:email).uniq
    unique_mentors = mentors.map(&:email).uniq

    mentee_match_percent = (100.0 * matched_mentees.size / unique_mentees.size)
    mentor_match_percent = (100.0 * matched_mentors.size / unique_mentors.size)

    mentees_to_mentors = matched_mentors_to_mentees.invert

    mentees_to_ranked_results = mentees_to_mentors.map do |mentee, mentor|
      mentee_preferences = mentees_to_preferences.fetch(mentee)
      # Use 1-based indexing to represent first pick is 1
      mentor_rank = T.must(mentee_preferences.find_index(mentor)) + 1
      [mentee, mentor_rank]
    end.to_h

    mentors_to_ranked_results = matched_mentors_to_mentees.map do |mentor, mentee|
      mentor_preferences = mentors_to_preferences.fetch(mentor)
      # Use 1-based indexing to represent first pick is 1
      mentee_rank = T.must(mentor_preferences.find_index(mentee)) + 1
      [mentor, mentee_rank]
    end.to_h

    puts("Median mentee # possible mentors = #{DescriptiveStatistics.median(mentees_to_preferences.values.map(&:size))}")
    puts("Median mentee paired mentor rank (e.g. 4 means 4th best) = #{DescriptiveStatistics.median(mentees_to_ranked_results.values)}")
    puts


    puts("Median mentor # possible mentees = #{DescriptiveStatistics.median(mentors_to_preferences.values.map(&:size))}")
    puts("Median mentor paired match rank (e.g. 4 means 4th best) = #{DescriptiveStatistics.median(mentors_to_ranked_results.values)}")
    puts

    puts("# Mentor / Mentee pairs in same state = #{matched_mentors_to_mentees.select {|mentor, mentee| mentor.state == mentee.state}.size} / #{matched_mentors_to_mentees.size}")
    puts("Median seniority difference = #{DescriptiveStatistics.median(matched_mentors_to_mentees.map{|mentor, mentee| mentor.seniority - mentee.seniority})}")
    puts

    puts("# mentors preferring IMG = " \
      "#{mentors.select(&:prefers_mentoring_international).size}")
      
    puts("# IMG mentees = " \
      "#{mentees.select(&:is_international).size}")
    
    puts("# IMG mentees paired w/ mentors preferring IMG = " \
    "#{matched_mentors_to_mentees.select {|mentor, mentee| mentor.prefers_mentoring_international && mentee.is_international}.size}")
    puts

    puts("# matched mentees: #{matched_mentees.count}")
    puts("# eligible mentees: #{unique_mentees.size}")
    puts("% mentee match: #{mentee_match_percent}%")
    unmatched_mentees = mentees.reject {|mentee| matched_mentees.include?(mentee.email)}
    puts("\nUnmatched mentees (#{unmatched_mentees.size}): #{unmatched_mentees.map(&:email).sort}\n\n")

    puts("# matched mentors: #{matched_mentors.size}")
    puts ("# eligible mentors: #{unique_mentors.size}")
    puts("% mentor match: #{mentor_match_percent}%")
    unmatched_mentors = mentors.reject {|mentor| matched_mentors.include?(mentor.email)}.group_by(&:email).map(&:first)
    puts("\nUnmatched mentors (#{unmatched_mentors.size}): #{unmatched_mentors.sort}\n\n")
  end

  # Takes input proposers and accepters as hash from IDs -> List of IDs in order of preference (descending)
  # and performs a basic Gale Shapley stable matching algorithm.
  #
  # Returns Hash mapping proposers -> acceptors.
  sig do
    params(
      proposers: T::Hash[Person2025, T::Array[Person2025]],
      acceptors: T::Hash[Person2025, T::Array[Person2025]],
    ).returns(T::Hash[Person2025, Person2025])
  end
  private_class_method def self.gale_shapley(
    proposers:,
    acceptors:
  )
    proposers = T.cast(Marshal.load(Marshal.dump(proposers)), T::Hash[Person2025, T::Array[Person2025]])
    acceptors = T.cast(Marshal.load(Marshal.dump(acceptors)), T::Hash[Person2025, T::Array[Person2025]])

    # Filter out any rankings of proposers/acceptors who did not reciprocally rank the acceptor/proposer.
    proposers = proposers.map {|proposer, preferences| [proposer, preferences.select {|acceptor| acceptors.fetch(acceptor).include?(proposer)}]}.to_h
    acceptors = acceptors.map {|acceptor, preferences| [acceptor, preferences.select {|proposer| proposers.fetch(proposer).include?(acceptor)}]}.to_h

    matches = T.let({}, T::Hash[Person2025, Person2025])

    proposer = T.let(pick_next_proposer(proposers: proposers, matches: matches), T.nilable(Person2025))

    # While we have an active proposer with options, we continue.
    while !proposer.nil?

      top_choice = get_and_remove_top_choice(proposer, proposers)

      existing_match = matches[top_choice]
      if existing_match.nil?
        matches[top_choice] = proposer
      else
        if prefers?(acceptor: top_choice, proposer1: proposer, proposer2: existing_match, acceptors: acceptors)
          matches[top_choice] = proposer
        end
      end

      proposer = pick_next_proposer(proposers: proposers, matches: matches)
    end

    matches
  end

  sig do
    params(
      proposers: T::Hash[Person2025, T::Array[Person2025]],
      matches: T::Hash[Person2025, Person2025]
    ).returns(T.nilable(Person2025))
  end
  private_class_method def self.pick_next_proposer(proposers:, matches:)
    with_choices = proposers
      # only select proposers who have potential acceptors left.
      .select {|proposer, choices| !choices.empty?}
      # only select proposers who haven't already been matched.
      .select {|proposer, _| !matches.values.include?(proposer)}

    with_choices.keys.first
  end

  sig do
    params(
      proposer: Person2025,
      proposers: T::Hash[Person2025, T::Array[Person2025]]
    ).returns(Person2025)
  end
  private_class_method def self.get_and_remove_top_choice(proposer, proposers)
    T.must(proposers.fetch(proposer).shift)
  end

  sig do
    params(
      acceptor: Person2025,
      proposer1: Person2025,
      proposer2: Person2025,
      acceptors: T::Hash[Person2025, T::Array[Person2025]]
    ).returns(T::Boolean)
  end
  private_class_method def self.prefers?(acceptor:, proposer1:, proposer2:, acceptors:)
    preferences = acceptors.fetch(acceptor)
    proposer1_preference = preferences.index(proposer1) || -1
    proposer2_preference = preferences.index(proposer2) || -1
    proposer1_preference < proposer2_preference
  end

end

File: src/mentor_matching/matching.py
from __future__ import annotations

from statistics import median

from mentor_matching.models import MatchStatistics, NAN, Person2025
from mentor_matching.preferences import Preferences


class Matching:
    @classmethod
    def multiplicity(cls, person: Person2025) -> list[Person2025]:
        if person.max_num_mentees <= 1:
            return [person]
        return [person.with_id(f"{person.id}{index}") for index in range(1, person.max_num_mentees + 1)]

    @classmethod
    def match(
        cls, people: list[Person2025], previously_matched: set[str] | None = None
    ) -> tuple[dict[Person2025, Person2025], MatchStatistics]:
        excluded_pairs = previously_matched or set()
        diagnostics: list[str] = []
        mentors = [mentor for person in people if person.is_mentor for mentor in cls.multiplicity(person)]
        mentees = [person for person in people if person.is_mentee]

        if excluded_pairs:
            diagnostics.append(
                f"Excluding {len(excluded_pairs)} previously-matched pair(s) from consideration"
            )

        mentees_to_preferences = Preferences.compute_mentee_to_mentor_preferences(
            mentees=mentees,
            mentors=mentors,
            previously_matched=excluded_pairs,
        )
        mentees_with_no_preferences = [
            mentee for mentee, preferences in mentees_to_preferences.items() if not preferences
        ]
        if mentees_with_no_preferences:
            diagnostics.append(
                "Filtering out "
                f"{len(mentees_with_no_preferences)} mentees with no preferences:\n"
                f"{mentees_with_no_preferences}"
            )
        mentees_to_preferences = {
            mentee: preferences for mentee, preferences in mentees_to_preferences.items() if preferences
        }
        diagnostics.append(f"Num mentees: {len({person.email for person in people if person.is_mentee})}")

        mentors_to_preferences = Preferences.compute_mentor_to_mentee_preferences(
            mentees=mentees,
            mentors=mentors,
            previously_matched=excluded_pairs,
        )
        mentors_with_no_preferences = [
            mentor for mentor, preferences in mentors_to_preferences.items() if not preferences
        ]
        if mentors_with_no_preferences:
            diagnostics.append(
                f"Filtering out {len(mentors_with_no_preferences)} mentors with no preferences:\n"
            )
        mentors_to_preferences = {
            mentor: preferences for mentor, preferences in mentors_to_preferences.items() if preferences
        }
        diagnostics.append(f"Num mentors: {len({person.email for person in people if person.is_mentor})}")
        diagnostics.append(
            "Num mentor slots (accounting for multiple mentees per mentor): "
            f"{len(mentors_to_preferences)}"
        )

        matched_mentors_to_mentees = cls._gale_shapley(
            proposers=mentees_to_preferences,
            acceptors=mentors_to_preferences,
        )

        return matched_mentors_to_mentees, cls._compute_match_statistics(
            matched_mentors_to_mentees=matched_mentors_to_mentees,
            mentees=mentees,
            mentors=mentors,
            mentees_to_preferences=mentees_to_preferences,
            mentors_to_preferences=mentors_to_preferences,
            diagnostics=diagnostics,
        )

    @classmethod
    def format_results(cls, matched_mentors_to_mentees: dict[Person2025, Person2025]) -> str:
        rows = sorted(
            f"{mentee.name};{mentee.email};{mentor.name};{mentor.email}"
            for mentor, mentee in matched_mentors_to_mentees.items()
        )
        return "\n".join(rows)

    @classmethod
    def _compute_match_statistics(
        cls,
        matched_mentors_to_mentees: dict[Person2025, Person2025],
        mentees: list[Person2025],
        mentors: list[Person2025],
        mentees_to_preferences: dict[Person2025, list[Person2025]],
        mentors_to_preferences: dict[Person2025, list[Person2025]],
        diagnostics: list[str],
    ) -> MatchStatistics:
        matched_mentee_emails = sorted({mentee.email for mentee in matched_mentors_to_mentees.values()})
        matched_mentor_emails = sorted({mentor.email for mentor in matched_mentors_to_mentees})
        unique_mentee_emails = sorted({mentee.email for mentee in mentees})
        unique_mentor_emails = sorted({mentor.email for mentor in mentors})

        mentees_to_mentors = {mentee: mentor for mentor, mentee in matched_mentors_to_mentees.items()}
        mentee_ranks = [
            mentees_to_preferences[mentee].index(mentor) + 1 for mentee, mentor in mentees_to_mentors.items()
        ]
        mentor_ranks = [
            mentors_to_preferences[mentor].index(mentee) + 1 for mentor, mentee in matched_mentors_to_mentees.items()
        ]

        unmatched_mentee_emails = sorted(
            mentee.email for mentee in mentees if mentee.email not in set(matched_mentee_emails)
        )
        unmatched_mentor_emails = sorted(
            {
                mentor.email
                for mentor in mentors
                if mentor.email not in set(matched_mentor_emails)
            }
        )

        same_state_pair_count = sum(
            1 for mentor, mentee in matched_mentors_to_mentees.items() if mentor.state == mentee.state
        )
        total_pair_count = len(matched_mentors_to_mentees)
        seniority_differences = [mentor.seniority - mentee.seniority for mentor, mentee in matched_mentors_to_mentees.items()]
        img_preferring_mentor_count = sum(1 for mentor in mentors if mentor.prefers_mentoring_international)
        img_mentee_count = sum(1 for mentee in mentees if mentee.is_international)
        img_preference_pair_count = sum(
            1
            for mentor, mentee in matched_mentors_to_mentees.items()
            if mentor.prefers_mentoring_international and mentee.is_international
        )

        return MatchStatistics(
            diagnostics=tuple(diagnostics),
            matched_mentee_emails=tuple(matched_mentee_emails),
            matched_mentor_emails=tuple(matched_mentor_emails),
            unmatched_mentee_emails=tuple(unmatched_mentee_emails),
            unmatched_mentor_emails=tuple(unmatched_mentor_emails),
            mentee_match_percent=cls._percent(len(matched_mentee_emails), len(unique_mentee_emails)),
            mentor_match_percent=cls._percent(len(matched_mentor_emails), len(unique_mentor_emails)),
            median_possible_mentors_for_mentee=cls._median_or_none(
                [len(value) for value in mentees_to_preferences.values()]
            ),
            median_possible_mentees_for_mentor=cls._median_or_none(
                [len(value) for value in mentors_to_preferences.values()]
            ),
            median_matched_mentor_rank_for_mentee=cls._median_or_none(mentee_ranks),
            median_matched_mentee_rank_for_mentor=cls._median_or_none(mentor_ranks),
            same_state_pair_count=same_state_pair_count,
            total_pair_count=total_pair_count,
            median_seniority_difference=cls._median_or_none(seniority_differences),
            img_preferring_mentor_count=img_preferring_mentor_count,
            img_mentee_count=img_mentee_count,
            img_preference_pair_count=img_preference_pair_count,
        )

    @classmethod
    def _gale_shapley(
        cls,
        proposers: dict[Person2025, list[Person2025]],
        acceptors: dict[Person2025, list[Person2025]],
    ) -> dict[Person2025, Person2025]:
        proposer_preferences = {proposer: list(preferences) for proposer, preferences in proposers.items()}
        acceptor_preferences = {acceptor: list(preferences) for acceptor, preferences in acceptors.items()}

        proposer_preferences = {
            proposer: [
                acceptor
                for acceptor in preferences
                if acceptor in acceptor_preferences and proposer in acceptor_preferences[acceptor]
            ]
            for proposer, preferences in proposer_preferences.items()
        }
        acceptor_preferences = {
            acceptor: [
                proposer
                for proposer in preferences
                if proposer in proposer_preferences and acceptor in proposer_preferences[proposer]
            ]
            for acceptor, preferences in acceptor_preferences.items()
        }

        matches: dict[Person2025, Person2025] = {}
        proposer = cls._pick_next_proposer(proposers=proposer_preferences, matches=matches)
        while proposer is not None:
            top_choice = cls._get_and_remove_top_choice(proposer=proposer, proposers=proposer_preferences)
            existing_match = matches.get(top_choice)
            if existing_match is None or cls._prefers(
                acceptor=top_choice,
                proposer1=proposer,
                proposer2=existing_match,
                acceptors=acceptor_preferences,
            ):
                matches[top_choice] = proposer
            proposer = cls._pick_next_proposer(proposers=proposer_preferences, matches=matches)

        return matches

    @staticmethod
    def _pick_next_proposer(
        proposers: dict[Person2025, list[Person2025]],
        matches: dict[Person2025, Person2025],
    ) -> Person2025 | None:
        matched_proposers = set(matches.values())
        for proposer, choices in proposers.items():
            if choices and proposer not in matched_proposers:
                return proposer
        return None

    @staticmethod
    def _get_and_remove_top_choice(
        proposer: Person2025,
        proposers: dict[Person2025, list[Person2025]],
    ) -> Person2025:
        choices = proposers[proposer]
        if not choices:
            raise RuntimeError(f"Proposer has no choices left: {proposer}")
        return choices.pop(0)

    @staticmethod
    def _prefers(
        acceptor: Person2025,
        proposer1: Person2025,
        proposer2: Person2025,
        acceptors: dict[Person2025, list[Person2025]],
    ) -> bool:
        preferences = acceptors[acceptor]
        proposer1_preference = preferences.index(proposer1) if proposer1 in preferences else -1
        proposer2_preference = preferences.index(proposer2) if proposer2 in preferences else -1
        return proposer1_preference < proposer2_preference

    @staticmethod
    def _median_or_none(values: list[int]) -> float | None:
        if not values:
            return None
        return float(median(values))

    @staticmethod
    def _percent(numerator: int, denominator: int) -> float:
        if denominator == 0:
            return NAN
        return 100.0 * numerator / denominator
